File: leetcode/swordToOffer/016/code.cpp
#include<iostream>
#include<unordered_map>

using namespace std;

class Solution {
public:
    // 哈希映射
int lengthOfLongestSubstring(string s) {
    int n = s.length();
    if (n <= 1) return n;

    int maxLen = 0;
    int left = 0, right = 0;
    // 存储窗口中每个字符及其位置的下一个位置
    // 表示如果再次碰到这个字符的时候，那么 left 就跳到这个字符所在位置的下一个位置
    unordered_map<char, int> window;
    while (right < n) {
        char currChar = s[right];
        unordered_map<char, int>::iterator it = window.find(currChar);
        int currCharIndex = (it == window.end() ? -1 : it->second);
        left = max(left, currCharIndex);
        maxLen = max(maxLen, right - left + 1);

        window[currChar] = right + 1;
        right++;
    }
    return maxLen;
}
};

int main()
{
    Solution solution;
    string b = "xyzabcasoyx";
    int result = solution.lengthOfLongestSubstring(b);
    cout<<result;
}

File: leetcode/swordToOffer/020/code.py
class Solution(object):
    def countSubstrings(self, s):
        result = 0
        left = 0
        right = 0
        if(len(s) <2):
            return len(s)
        for i in range(len(s)-1):
            if(s[i+1] == s[i]):               # 偶数回文子串
                left = i
                right = i+1
                while(s[left]==s[right]):   # 相等
                    result = result + 1     
                    left = left -1
                    right = right + 1
                    if(left < 0  or right > len(s)-1):
                        break
            left = i
            right = i
            while(s[left]==s[right]):
                result = result + 1
                left = left -1
                right = right + 1
                if(left < 0  or right > len(s)-1):
                    break
        
        return (result + 1)

b = Solution()
print(b.countSubstrings("abacaba"))